### PR TITLE
xds: Envoy proto sync to 2022-04-08

### PR DESCRIPTION
**Pull request by sergiitk**
### Proto updates

1. `cncf/xds`: Sort xds/import.sh protos alphabetically
2. `cncf/xds`: Sync protos to https://github.com/cncf/xds/commit/d92e9ce0af512a73a3a126b32fa4920bee12e180 (commit 2021-12-16, corresponding to envoy cl/440193522). It's a no-op for used protos, but helpful to import the latest matcher.proto
3. `cncf/xds`: Import xds/type/matcher/v3/matcher.proto with dependencies
4. `envoyproxy/protoc-gen-validate`: Sync protos to https://github.com/envoyproxy/protoc-gen-validate/commit/dfcdc5ea103dda467963fb7079e4df28debcfd28 (commit 2022-03-10, corresponding to envoy cl/440193522) to pick up ignore_empty field required for the following envoy sync
5. `envoyproxy/envoy` Sync protos to https://github.com/envoyproxy/envoy/commit/e33f444b4c5ee783a8a1126f3186832f4a031a39 (commit 2022-04-07, cl/440193522). This is the minimal version needed to pick up `ClusterSpecifierPlugin.is_optional`.
   a. Generated code: `AggregatedDiscoveryServiceGrpc` was regenerated from the updated proto. This is a no-op, just a minor change to the docblocks.
   b. Deprecated fields had to be taken care of manually, see "Manual updates to the code" below.
6. `envoyproxy/envoy` Sync protos to the latest imported version https://github.com/envoyproxy/envoy/commit/5d74719102f461bc57e85acdda706e0a8df9b12d (commit 2022-04-08, cl/443359189). Not needed for anything specific, just the last version, and was easy to import.

### Manual updates to the code

As the result of https://github.com/envoyproxy/envoy/commit/e33f444b4c5ee783a8a1126f3186832f4a031a39 sync:

1. Deprecated `ConfigSource.path` replaced with the `ConfigSource.path_config_source` in test fake resources. The `ConfigSource.path` isn't in active code paths, so no prod code changes needed.
2. Suppress `CertificateValidationContext.match_subject_alt_names` deprecations in test files. Surprisingly, we don't report deprecations in prod files, despite the fact this field is used in prod code a few times.

### CODE REVIEW

**Comment by sergiitk** [comment]
@sanjaypujare FYI field `CertificateValidationContext.match_subject_alt_names` got deprecated. Seems to be relevant to some of the work you did: https://github.com/grpc/grpc-java/blame/8e65700edc62168fdc85b5b20a298f0280aa23c1/xds/src/main/java/io/grpc/xds/ClientXdsClient.java#L602

**Comment by sanjaypujare** [comment]
> @sanjaypujare FYI field `CertificateValidationContext.match_subject_alt_names` got deprecated. Seems to be relevant to some of the work you did: https://github.com/grpc/grpc-java/blame/8e65700edc62168fdc85b5b20a298f0280aa23c1/xds/src/main/java/io/grpc/xds/ClientXdsClient.java#L602

Correct but unless Traffic Director stops using the deprecated fields we cannot make the change unilaterally on our side

**Comment by sergiitk** [comment]
@sanjaypujare Yes, just a heads up. What we could do is to add support for the new field, while keeping the deprecated field. Up to you though, I just wanted to inform you of the deprecation.

**Comment by sanjaypujare** [comment]
> ... add support for the new field, while keeping the deprecated field. ...

Already done! Although not tested with a real xDS control plane since we don't have any to test with

**Comment by sergiitk** [comment]
Wait, with what field did it get replaced? I tried to update tests to replace it with `match_typed_subject_alt_names`  in unit tests, but tests don't pass with it.

**Comment by sanjaypujare** [comment]
> Wait, with what field did it get replaced? I tried to update tests to replace it with `match_typed_subject_alt_names` in unit tests, but tests don't pass with it.

I may have misunderstood or misspoke. May be the subject_alt_name needs to be implemented because it's a recent addition. Good point. BTW all other languages also need the change

**Comment by sergiitk** [comment]
Ah, glad we figured it out. Let me know if there's anything I can do to help with the review. Going through each commit is probably the easiest.

**Comment by sanjaypujare** [comment]
> I tried to update tests to replace it with match_typed_subject_alt_names in unit tests, 

BTW, I don't think you should remove or update any tests using deprecated fields just yet. We depend on these tests to ensure interoperability with other components (e.g. Traffic Director) until we verify that the deprecated fields have been completely replaced by these other components.

File: xds/third_party/xds/import.sh [hunk 1]
```diff
@@ -26,26 +26,26 @@ TARGET_PROTO_BASE_DIR=src/main/proto
...
-xds/annotations/v3/sensitive.proto
 udpa/annotations/status.proto
-xds/annotations/v3/status.proto
 udpa/annotations/versioning.proto
-xds/annotations/v3/versioning.proto
-xds/data/orca/v3/orca_load_report.proto
-xds/service/orca/v3/orca.proto
 udpa/type/v1/typed_struct.proto
-xds/type/v3/typed_struct.proto
+xds/annotations/v3/migrate.proto
```
Comment (sanjaypujare) [inline xds/third_party/xds/import.sh:35]: This comment is for the whole commit "Sort xds/import.sh protos alphabetically". 

- have we decided that the strict alphabetical order is better than for example having related protos together? e.g. the 2 versions of migrate.proto together?
- we will eventually be dropping support for v2 xds/protos? If it is imminent is this reordering worthwhile?

Comment (sergiitk) [reply xds/third_party/xds/import.sh:35]: 1. Yes, there's even a comment on top of the import, same as with all other `import.sh` files. We just didn't do it in xds/import.sh for some reason. https://github.com/grpc/grpc-java/blob/538db03d56630df042c44c2e81363be924a2be8f/xds/third_party/xds/import.sh#L26
2. Don't know if or when we'll drop them. Reordering was cheap for me - select lines -> ctrl+F5 to sort. Didn't consider this to be a price.

File: xds/third_party/xds/import.sh [hunk 2]
```diff
@@ -41,10 +41,14 @@ xds/annotations/v3/versioning.proto
 xds/core/v3/authority.proto
 xds/core/v3/collection_entry.proto
 xds/core/v3/context_params.proto
+xds/core/v3/extension.proto
```
Comment (sanjaypujare) [inline xds/third_party/xds/import.sh:44]: `xds/third_party/envoy/import.sh` also imports an `extension.proto` but I can see that they are different although `TypedExtensionConfig` definition is identical so looks like this one is ultimately meant to replace the other one. Just pointing out and most probably not a problem and something we are forced to live with ...

Comment (sergiitk) [reply xds/third_party/xds/import.sh:44]: Interesting find, thank you. I checked in the envoy repo; they don't use `TypedExtensionConfig` from cncf/xds `extension.proto` yet. I don't think there's anything we can do until envoy does the migration.
Regarding matchers - AFAIK they're slightly different, unfortunately. I believe there will be unification, and the migration to cncf/xds.

**Review by sanjaypujare — APPROVED** [review]
LGTM

**Pull request merged**